Add io.github.piotrek_k._2dTaskBoard

piotrek-k opened the pull request:
<!-- ⚠️⚠️ Submission pull request MUST be made against the `new-pr` **base branch** ⚠️⚠️  -->

<!-- 💡 Go to the preview tab to click the links below 💡 -->

### Please confirm your submission meets all the criteria

<!-- 💡 Please replace each `[ ]` with `[X]` when the step is complete 💡 -->

- [x] Please describe the application briefly. <!-- insert the description here -->
    -  Task management productivity app with markdown support. Tasks can be moved between rows and columns. Users data is being stored in the filesystem as files.
- [x] The domain used for the application ID is [controlled by the application developer(s)][appid-domain] and the [application id guidelines][appid] are followed.
    - It's hosted on github, so I've used io.github domain
- [x] I have read and followed all the [Submission requirements][reqs] and the [Submission guide][reqs2].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link:**

<!-- 💡 Mention any additional maintainers needed below 💡 -->

<!-- ⚠️⚠️ DO NOT modify anything below this line ⚠️⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[reqs2]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.desktop:
this should be (submitted) upstream

hfiguiere replied:
it's also never referenced nor installed

piotrek-k replied:
ok, moved it to https://github.com/piotrek-k/2dTaskBoard/blob/master/desktop/flathub/io.github.piotrek_k._2dTaskBoard.desktop

I've also added `install` command in `io.github.piotrek_k._2dTaskBoard.yml`, I'm not sure if that's what you were referring to

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.metainfo.xml:
this should be (submitted) upstream

hfiguiere replied:
it's also never referenced nor installed

piotrek-k replied:
Moved to upstream, added to io.github.piotrek_k._2dTaskBoard.yml

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.metainfo.xml:
```diff
@@ -0,0 +1,59 @@
+        </screenshot>
+    </screenshots>
+
+    <icon type="remote">https://raw.githubusercontent.com/piotrek-k/flathub/ab0f2c00731c68f33d9f61f1d813b6e9e9c44101/icons/icon.svg</icon>
```
```suggestion
```

piotrek-k replied:
removed

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.metainfo.xml:
```diff
@@ -0,0 +1,59 @@
+
+    <screenshots>
+        <screenshot type="default">
+            <image>https://raw.githubusercontent.com/piotrek-k/flathub/refs/heads/my-app-submission/screenshots/board_view.png</image>
```
screenshot URL should never reference the tip of a branch. It's not immutable.

piotrek-k replied:
Now it's referring to commit, let me know if this is what you meant

hfiguiere commented on icons/icon.svg:
this should be (submitted) upstream

hfiguiere replied:
it's also never referenced nor installed

piotrek-k replied:
renamed, moved to upstream, added to `io.github.piotrek_k._2dTaskBoard.yml`

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.metainfo.xml:
```diff
@@ -0,0 +1,59 @@
+
+    <releases>
+        <release version="0.4.1" date="2025-03-15">
+            <description></description>
```
```suggestion
```

piotrek-k replied:
removed

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,56 @@
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --filesystem=home
```
why does it need full home access?

piotrek-k replied:
This application stores data as files in a file system (attachments, task contents). I want to allow users to track changes using git or synchronise their tasks using a cloud provider. 

The application works without this permission, but without it, the File Chooser will not display directories that the user is familiar with, which can be confusing.

bbhtt replied:
> The application works without this permission, but without it, the File Chooser will not display directories that the user is familiar with, which can be confusing.

Why does it not use portals for the filechooser?

piotrek-k replied:
Is using this permission strictly forbidden?

My app was initially meant to work in web browser and managing files has been done via browser's FileSystem API. What I've done here is I've turned it into Electron app, which allows me to distribute it as fully offline desktop app.

Using this FileSystem API already limits what I can access on users device, web browser requires user to point to the directory I can operate in.

I'm not yet sure what those portals are, but seems like some special API for accessing filesystem? If yes, then it defeats what I wanted to accomplish here: single codebase between web and desktop

bbhtt replied:
> I'm not yet sure what those portals are

https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.FileChooser.html

> The FileChooser portal allows sandboxed applications to ask the user for access to files outside the sandbox. The portal backend will present the user with a file chooser dialog.
> The selected files will be made accessible to the application (which may involve adding it to the [Documents portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Documents.html#org-freedesktop-portal-documents)).

Electron supports it.

piotrek-k replied:
I'm afraid I still don't understand.
My app consists of basically a browser that runs a web app. Everything that happens inside my app, happens through browser's api.

I couldn't find any official documentation for it, but I've found an issue in Chromium bugtracker that suggests that it might use XDG filechooser already:
https://issues.chromium.org/issues/41414524

> Implement support for the XDG file chooser portal
> This allows use on Linux of the XDG portal for file selection rather
than directly calling out to the GTK/KDE dialogs. 

Now, what happens in my app when I remove this permission?
```
- --filesystem=home
```

It shows filepicker:
![image](https://github.com/user-attachments/assets/67d290f0-b16d-41c2-b1f4-4fa86631369f)

but there are no files displayed. I can create a directory, but when I close my app and open it again, I cannot see this directory anymore.

bbhtt replied:
> but there are no files displayed. I can create a directory, but when I close my app and open it again, I cannot see this directory anymore.

Yes if it was using portal there it'd show all files without the permission. What version of electron are you using?

piotrek-k replied:
You were right. I've updated electron to the newest version and it now works correctly without this permission. So I've removed `--filesystem=home` from io.github.piotrek_k._2dTaskBoard.yml

piotrek-k commented on io.github.piotrek_k._2dTaskBoard.desktop:
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Version=0.4.1
```
I've removed it completly
https://github.com/piotrek-k/2dTaskBoard/blob/master/desktop/flathub/io.github.piotrek_k._2dTaskBoard.desktop

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,61 @@
+app-id: io.github.piotrek_k._2dTaskBoard
```
The id should be `io.github.piotrek_k.2dTaskBoard`

> The domain portion must be in lowercase and must convert dash - to underscore _ and also prefix any intial digits with an underscore _.

piotrek-k replied:
It was like that before, but I've ran a validator and it said:

> E: io.github.piotrek_k.2dTaskBoard:3: cid-has-number-prefix
>      io.github.piotrek_k.2dTaskBoard: 2dTaskBoard → _2dTaskBoard
>    The component ID contains a segment starting with a number. Starting a segment of the
>    reverse-DNS ID with a number is strongly discouraged, to keep interoperability with other tools
>    such as D-Bus. Ideally, prefix these segments with an underscore.

I used this command:
```
flatpak run --command=flatpak-builder-lint org.flatpak.Builder appstream io.github.piotrek_k._2dTaskBoard.metainfo.xml
```

hfiguiere replied:
looks like there a contradiction somewhere....

bbhtt replied:
Keep it as is, I gave you an exception for appid-url-not-reachable https://github.com/flathub-infra/flatpak-builder-lint/commit/31a3c4537b721c1b3e8220f2e1ec030e9fd10f2b

piotrek-k replied:
thanks

hfiguiere commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,61 @@
+    sources:
+      - type: git
+        url: https://github.com/piotrek-k/2dTaskBoard.git
+        tag: v0.4.1
```
Please add the commit for the tag as well.

piotrek-k replied:
Sure, commit id added

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

bbhtt commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,62 @@
+    subdir: main
+    sources:
+      - type: git
+        url: https://github.com/piotrek-k/2dTaskBoard.git
```
the upstream repo has no description readme, instructions, or explains what it is, I'm unlikely to accept in this state https://github.com/piotrek-k/2dTaskBoard

what is the web and desktop parts?

For the desktop I see a single file that opens something in localhost https://github.com/piotrek-k/2dTaskBoard/blob/53bd77353cdb2984b39a551414565b8503af8d53/desktop/index.js#L69

how is that an app?

piotrek-k replied:
> the upstream repo has no description readme

I've just added it https://github.com/piotrek-k/2dTaskBoard

> what is the web and desktop parts?

As I said before, 2dTaskBoard is a web app, packed into Electron.
Building this app looks like that:
1. Building web part (html + css + js files are created as a result)
2. Copying built web app inside `desktop` directory
3. Building Electron app

The only thing Electron as of now does, is it creates local webserver and hosts my web app inside app window

piotrek-k commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,62 @@
+  env:
+    NPM_CONFIG_LOGLEVEL: info
+modules:
+  - name: electron-sample-app
```
Changed `electron-sample-app` to `2dTaskBoard`

piotrek-k commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,62 @@
+      env:
+        XDG_CACHE_HOME: /run/build/electron-sample-app/flatpak-node/cache
+        npm_config_cache: /run/build/electron-sample-app/flatpak-node/npm-cache
+        npm_config_offline: 'true'
```
Changed `electron-sample-app` to `2dTaskBoard`

bbhtt commented:
why is the first run screen always starting with this confusing message?

![image](https://github.com/user-attachments/assets/d0680553-8f17-423d-9067-99963eab374b)


I ran it for the first time, it didn't have "directory for storing tasks".

bbhtt commented:
why is this empty? 

![image](https://github.com/user-attachments/assets/559b970e-c420-4303-a46b-0b9953c3f841)

piotrek-k commented:
I've modified it a bit, an app should now display more descriptive message at startup:

![image](https://github.com/user-attachments/assets/be6cd16c-7a53-4254-9d1a-bdbc151edc97)

I've also removed some options from top menu that are not useful at the moment.

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

bbhtt commented:
It says we lost access to the directory where tasks were stored everytime from the second time.

hfiguiere commented:
I'm gonna take a guess: it needs a `persist` for the directory it stores things.

piotrek-k commented:
> It says "we lost access to the directory where tasks were stored" everytime from the second time.

It seems like it's a known Electron issue https://github.com/electron/electron/issues/41957 Chrome support persistent permissions but it's not implemented in Electron yet it seems.

My app does indeed remember last directory that user has chosen and tries to restore it (that's why a message says "We have **lost** access to the directory..."), but as I see from logs it requires "user gesture" to restore permissions. Notice that it is enough that the user clicks on the button, he need not to choose directory again, because after clicking it's being loaded automatically (as it is a "gesture" that Chromium requires).

> I'm gonna take a guess: it needs a persist for the directory it stores things.

Do you mean this: https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access
Can you give a hint on how can I use it in my case? It looks like it lets me map directory from user's filesystem to filesystem inside my app's sandbox. But how can it solve my issue? Electron is probably still going to request user gesture

bbhtt commented:
Try setting a default directory to something like `xdg-documents/2dTaskBoard` in the app and give `--filesystem=xdg-documents/2dTaskBoard` permission.

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

piotrek-k commented:
There were some minor UI fixes, so I decided to update an app to the next version: 0.4.3

Coming back to our discussion about directory persistence. I've researched it a little more and I still believe [it is not possible right now with the latest Electron version](https://stackoverflow.com/questions/79544391/how-can-i-preserve-filesystem-api-access-to-a-directory-between-app-launches), so it still requires user to "perform a gesture" (in this case: click a button). But I tweaked a prompt message a little bit, so it does not look like an error anymore:
![image](https://github.com/user-attachments/assets/71d2677f-8a48-4794-9a21-bd11e8a6e06e)

piotrek-k commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,62 @@
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --device=dri
```
I needed to add this permission, because after I installed the build provided today by flathubbot, an app has been showing blank screen. After I added this permission it started to work again.

bbhtt replied:
it's fine

bbhtt commented on io.github.piotrek_k._2dTaskBoard.yml:
```diff
@@ -0,0 +1,62 @@
+        dest: main
+      - generated-sources-desktop.json
+      - generated-sources-web.json
+      - type: script
+        dest-filename: run.sh
+        commands:
+          - zypak-wrapper.sh /app/main/task-board-desktop "$@"
```
desktop file is not using this wrapper so launching via that does not work. use `desktop-file-edit` to change it here.

piotrek-k replied:
desktop file seems to be working now, here is what I've changed
https://github.com/piotrek-k/2dTaskBoard/commit/1b0fd4aeeb38fed06cf38c6588a4fc3f05d0ca01

piotrek-k commented:
bot, build io.github.piotrek_k._2dTaskBoard

bbhtt commented:
/merge
